queries/ecma/highlights.scm

```diff
@@ -24,10 +24,15 @@
  (#lua-match? @type "^[A-Z]"))
 
 ((identifier) @constant
- (#lua-match? @constant "^[A-Z_][A-Z%d_]+$"))
+ (#lua-match? @constant "^[A-Z_][A-Z%d_]*$"))
```

Comment:
`_` will be highlighted as a const, even though it is used to fill parameters only

Reply:
check now

Reply:
and now `__A` is broken.

it was working fine before

Reply:
`const A = 2`, A wouldn't match

I can fix that, oops

Reply:
Here's the fix..
![image](https://user-images.githubusercontent.com/29718261/222863363-22feb439-0047-4d50-9aac-10a313d95b5e.png)

Reply:
Think this new pattern still fails with `_1`, while the old one doesn't (but should `_1` be accepted)?

Reply:
Is _1 a legal identifier in js/ts?

Reply:
(also, think you should just add them into a testfile in ecma folder, just for the double check)

Reply:
> Is _1 a legal identifier in js/ts? And if it is, idk, that doesn't seem like a constant to me :)

It's a valid identifier, not an usual one

Reply:
lmk what you think

Reply:
> > Is _1 a legal identifier in js/ts? And if it is, idk, that doesn't seem like a constant to me :)
> 
> It's a valid identifier, not an usual one

Yeah I realized it's an identifier, sure, I mean to ask is it convention for such an identifier to be a constant? e.g. normally SCREAMING_SNAKE_CASE is for constants, but _1 doesn't really fit that

Reply:
tests look fine.

i dont think i should have the call on how `_1` should be, will let you and others decide

Reply:
You should take a look at this.

https://github.com/olimorris/onedarkpro.nvim/pull/164#issuecomment-1453912862

Reply:
@jNaimXIII I did mention this yesterday
https://github.com/nvim-treesitter/nvim-treesitter/issues/4434#issuecomment-1453559966

You just need to add a `@readonly` highlight group and i think it's fine

Reply:
> tests look fine.
> 
> i dont think i should have the call on how `_1` should be, will let you and others decide

In JS we usually use the `_` to denote throwaway parameters or, to introduce a variable with the same name in nested scope without loosing reference to the outer scope one.

<img width="880" alt="image" src="https://user-images.githubusercontent.com/115424698/222868054-ed219a79-03f8-4c13-aec6-1c377130565c.png">

So, if that's the case, the name of the variable shouldn't matter. Anything declared with a `const` should be highlighted the same as a constant.

Reply:
i think this is better to be delegated to semantic tokens highlighting in nvim 0.9 tbh

for this to highlight properly with treesitter is too costly

Reply:
The `const (smth)` highlighted as const part* is to be delegated to semantic tokens highlighting in 0.9, to be more clear

Reply:
So, not something tree-sitter should be doing on its own is what you're saying. Well then, when Neovim does ship this feature, themes will need to update their highlighting for this to take effect, right?

Reply:
`@readonly` is not a token linked to a default hlgroup by default in nvim 0.9, so theme maintainers need to add it in yes

Reply:
> @jNaimXIII I did mention this yesterday [#4434 (comment)](https://github.com/nvim-treesitter/nvim-treesitter/issues/4434#issuecomment-1453559966)
> 
> You just need to add a `@readonly` highlight group and i think it's fine

Why? It's fine to distinguish constants that are declared as constants and then leave their usage whatever it may be (in most cases, `@variable`). This is partially what using local definitions (locals.scm) fixes, except that it uses an inefficient and outdated method. 

In most languages, you'd have enum member/const declarations be highlighted as constant, but then their usage isn't exactly clear and relies on either `@variable` or regex/lua matching. So for the time being, this distinction here is fine until we get some better semantic token-like highlighting by default

Reply:
Just for context, I proposed this highlighting changed based on a VSCode theme called OneDarkPro. This theme highlights JS / TS code in this manner so, if you want you can refer to this.

https://github.com/Binaryify/OneDark-Pro

Reply:
> Just for context, I proposed this highlighting changed based on a VSCode theme called OneDarkPro. This theme highlights JS / TS code in this manner so, if you want you can refer to this.
> 
> https://github.com/Binaryify/OneDark-Pro

As I mentioned here, https://github.com/olimorris/onedarkpro.nvim/pull/164#issuecomment-1454005720 the theme is just marking every `constant` value with that color it seems. Then again, here comes that **semantic token** part which I don't understand much about.

Reply:
It's not really about the theme, but more what is captured under `@constant` queries and const variable definition identifiers weren't, but in this PR they are

Reply:
im pretty sure this is the behavior you want @jNaimXIII ?
![image](https://user-images.githubusercontent.com/29790821/222868667-4aac71bb-336b-4d1e-9cf3-eb2357f765f4.png)

Reply:
But the question still remains, does highlighting only the declaration to a different color while follow usage of the variable stays the same color make it more or less confusing

Reply:
@jNaimXIII semantic tokens is LSP stuffs, not treesitter stuffs, lets just keep it short

Reply:
Hence the issue mentioned here. https://github.com/olimorris/onedarkpro.nvim/pull/164#issuecomment-1453912862

Reply:
> Then again, here comes that semantic token part which I don't understand much about.

Given a definition marked say, `@definition.constant`, all subsequent identifiers that are the same as the definition identifier ini the same **scope** will be highlighted as `@constant`, because it is being referenced there. So, this relies on a good and accurate `locals.scm` file, where the proper rules are put under `@scope`, `@reference`, and `@definition.*`. It is not implemented the best in terms of performance, hence why it's disabled by default iirc.

Reply:
it will be very costly if only done from queries @jNaimXIII to have the behavior you expected.

it's better to wait for 0.9 and this can be much easier supported by theme maintainers

Reply:
Welp, playing the waiting game it is then. Thanks for discussing this though.

Reply:
> it will be very costly if only done from queries @jNaimXIII to have the behavior you expected.
> 
> it's better to wait for 0.9 and this can be much easier supported by theme maintainers 
> 

Do you mean this PR or using locals.scm is costly?

Reply:
> Do you mean this PR or using locals.scm is costly?

I mean by the amount of queries, and the possible tree range need to be parsed for each one you may need to do to have close to the same behavior he expected

what if the variable is from a different file, is a object key, ....
there is a lot to be done

Reply:
example is if someone do
```ts
const x = {foo: "bar"} as const;
```
can you really highlight `x.foo` with `foo` being readonly with treesitter with a short and performant query? I doubt it but if it's possible then nice.
not to mention nested ones :)

Reply:
> 

Now that you mention it, the highlighting isn't consistent on VSCode either.

<img width="711" alt="image" src="https://user-images.githubusercontent.com/115424698/222869395-59a2882c-7251-4741-8864-df678ebfae49.png">

<img width="677" alt="image" src="https://user-images.githubusercontent.com/115424698/222869411-db8cec3a-2cf7-423f-9d89-95e2c5ed1f53.png">

Reply:
> what if the variable is from a different file, is a object key, ....

Different file is not tree-sitters job so this shouldn't even be considered, tree-sitter is not an LSP.
By object key, can you provide a code example so I can understand what you mean better?

Reply:
> example is if someone do
> 
> ```ts
> const x = {foo: "bar"} as const;
> ```
> 
> can you really highlight `x.foo` with `foo` being readonly with treesitter with a short and performant query? I doubt it but if it's possible then nice. not to mention nested ones :)

I don't know js/ts all that well, but is this saying the object is a constant object?

Reply:
> I don't know js/ts all that well, but is this saying the object is a constant object?

yes

Reply:
You can use this as reference I guess.

<img width="584" alt="image" src="https://user-images.githubusercontent.com/115424698/222869639-c47c3932-79f5-4de6-9dc6-2939090292e1.png">

Reply:
How I'd highlight it from tree-sitter's POV is as follows:

x <- `@constant`
. <- `@punctuation.delimiter`
foo <- `@field`

foo is still a field, whether it's constant or not doesn't mean much in terms of highlighting (at least imo, maybe it means a lot to js/ts devs), that'd just be awkward imo; But to reiterate, this is more an LSPs job anyways

Reply:
well yes, that was my question since the start of this discussion
will (partial) semantic tokens highlighting using treesitter make reading the code more or less confusing

Reply:
> You can use this as reference I guess.
> 
> <img alt="image" width="584" src="https://user-images.githubusercontent.com/115424698/222869639-c47c3932-79f5-4de6-9dc6-2939090292e1.png">

Random note: I'm surprised `@variable.builtin` isn't highlighted differently (like console)

So, in the current state of things, `const object` would have object be highlighted as `@constant`, but subsequent references not. locals.scm adds subsequent scoped references to be highlighted `@constant` too, but at a performance cost

Reply:
> will (partial) semantic tokens highlighting using treesitter make reading the code more or less confusing

imo no, but this is language-dependent

Reply:
> well yes, that was my question since the start of this discussion will (partial) semantic tokens highlighting using treesitter make reading the code more or less confusing

Take a look at the following code. It won't matter if it's partially highlighted as long as you can differ between a `const` and `let` / `parameter`. You can instantly see that `firebase` and `settings` are constants whereas `doc` is function param.

<img width="518" alt="image" src="https://user-images.githubusercontent.com/115424698/222874633-09d3e8f5-a152-4af1-bac9-bcaf2448f050.png">